packages/core/shared/read-data-stream.ts

```diff
@@ -0,0 +1,69 @@
+import { StreamPartType, parseStreamPart } from './stream-parts';
+
+const NEWLINE = '\n'.charCodeAt(0);
+
+// concatenates all the chunks into a single Uint8Array
+function concatChunks(chunks: Uint8Array[], totalLength: number) {
+  const concatenatedChunks = new Uint8Array(totalLength);
+
+  let offset = 0;
+  for (const chunk of chunks) {
+    concatenatedChunks.set(chunk, offset);
+    offset += chunk.length;
+  }
+  chunks.length = 0;
+
+  return concatenatedChunks;
+}
+
+export async function* readDataStream(
+  reader: ReadableStreamDefaultReader<Uint8Array>,
+  {
+    isAborted,
+  }: {
+    isAborted?: () => boolean;
+  } = {},
+): AsyncGenerator<StreamPartType> {
```

Comment:
Now all paths ( streamingReactResponse, useChat, useCompletion, useAssistant ) go through this edge-safe stream parsing.